core/variant/required_ptr.h

```diff
@@ -0,0 +1,258 @@
+			_value(const_cast<std::remove_const_t<T_Other> *>(p_ptr)) {}
+	template <typename T_Other, std::enable_if_t<std::is_base_of_v<T, T_Other>, int> = 0>
+	_FORCE_INLINE_ RequiredParam &operator=(const T_Other *p_ptr) {
+		_value = const_cast<std::remove_const_t<T_Other> *>(p_ptr);
+		return *this;
+	}
+
+	template <typename T_Other, std::enable_if_t<std::is_base_of_v<T, T_Other>, int> = 0>
+	_FORCE_INLINE_ RequiredParam(const Ref<T_Other> &p_ref) :
+			_value(p_ref) {}
+	template <typename T_Other, std::enable_if_t<std::is_base_of_v<T, T_Other>, int> = 0>
+	_FORCE_INLINE_ RequiredParam &operator=(const Ref<T_Other> &p_ref) {
+		_value = p_ref;
+		return *this;
+	}
+
+	template <typename T_Other, std::enable_if_t<std::is_base_of_v<T, T_Other>, int> = 0>
+	_FORCE_INLINE_ RequiredParam(Ref<T_Other> &&p_ref) :
+			_value(std::move(p_ref)) {}
+	template <typename T_Other, std::enable_if_t<std::is_base_of_v<T, T_Other>, int> = 0>
+	_FORCE_INLINE_ RequiredParam &operator=(Ref<T_Other> &&p_ref) {
+		_value = std::move(p_ref);
+		return &this;
+	}
+
+	_FORCE_INLINE_ RequiredParam(const Variant &p_variant) :
+			_value(static_cast<T *>(p_variant.get_validated_object())) {}
+	_FORCE_INLINE_ RequiredParam &operator=(const Variant &p_variant) {
+		_value = static_cast<T *>(p_variant.get_validated_object());
+		return *this;
+	}
+};
+
+#define TMPL_EXTRACT_REQUIRED_PARAM_OR_FAIL(m_name, m_param, m_retval, m_msg, m_editor)                                        \
+	GODOT_DEPRECATED_BEGIN                                                                                                     \
+	if (unlikely(m_param._is_null())) {                                                                                        \
+		_err_print_error(FUNCTION_STR, __FILE__, __LINE__, "Required object \"" _STR(m_param) "\" is null.", m_msg, m_editor); \
+		return m_retval;                                                                                                       \
+	}                                                                                                                          \
+	std::conditional_t<                                                                                                        \
+			std::is_base_of_v<RefCounted, std::decay_t<decltype(m_param)>::element_type>,                                      \
+			Ref<std::decay_t<decltype(m_param)>::element_type>,                                                                \
+			std::decay_t<decltype(m_param)>::element_type *>                                                                   \
+			m_name = m_param._internal_ptr();                                                                                  \
+	GODOT_DEPRECATED_END                                                                                                       \
+	static_assert(true)
+
+#define EXTRACT_REQUIRED_PARAM_OR_FAIL(m_name, m_param) TMPL_EXTRACT_REQUIRED_PARAM_OR_FAIL(m_name, m_param, void(), "", false)
```

Comment:
This could use a comment that these are equivalent to the `ERR_FAIL` macros, just with required params.